### PR TITLE
Owasp middleware

### PR DESCRIPTION
**Pull request by yshchohaleu**

### CODE REVIEW

File: src/WaCore.Web/Middleware/SecureHeaders/Models/WaSecureHeadersMiddlewareConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+namespace WaCore.Web.Middleware.SecureHeaders.Models
+{
+    public class WaSecureHeadersMiddlewareConfiguration : ISecureHeadersMiddlewareConfiguration
```
Comment (HerbertBodner) [inline src/WaCore.Web/Middleware/SecureHeaders/Models/WaSecureHeadersMiddlewareConfiguration.cs]: Small typo: we use the prefix 'Wac' everywhere else, not 'Wa'

**Comment by HerbertBodner** [comment]
You were probably following this guidelines from the owasp project: https://www.owasp.org/index.php/OWASP_Secure_Headers_Project

Is there a specific reason, why you didn't implement X-Content-Type-Options and X-Permitted-Cross-Domain-Policies?

File: src/WaCore.Web/Middleware/SecureHeaders/SecureHeadersMiddlewareExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.AspNetCore.Builder;
+using WaCore.Web.Middleware.SecureHeaders.Models;
+
+namespace WaCore.Web.Middleware.SecureHeaders
```
Comment (rurku) [inline src/WaCore.Web/Middleware/SecureHeaders/SecureHeadersMiddlewareExtensions.cs]: It appears to be a common practice to put IApplicationBuilder extension methods in namespace Microsoft.AspNetCore.Builder so people can use them without adding a `using`

File: src/WaCore.Web/Middleware/SecureHeaders/WacSecureHeadersMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            {
+                httpContext.Response.Headers.Add(headerName, headerValue);
+            }
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            if (_config == null)
+            {
+                throw new ArgumentException($@"Expected an instance of the
```
Comment (rurku) [inline src/WaCore.Web/Middleware/SecureHeaders/WacSecureHeadersMiddleware.cs]: ArgumentNullException would be suitable here if you moved the check to the constructor

**Comment by yshchohaleu** [comment]
@HerbertBodner 
X-Permitted-Cross-Domain-Policies - as I understand this header is used only with Flash content. So I don't think it's necessary to include it.
X-Content-Type-Options - this one yes, I missed it.

File: src/WaCore.Sample.Middlewares/secureheaders.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿{
+    "WaSecureHeadersMiddlewareConfiguration": {
+        "HstsConfiguration": {
+            "MaxAge": 42,
+            "IncludeSubDomains": "true"
+        },
+        "HPKPConfiguration": {
+            "PinSha256": [
+                "e927fad33f9eb96126896413502a1034be0ca379dec377fb891feb9ebc720e47"
```
[the remark this answers is not in the document]
Comment (yshchohaleu) [reply src/WaCore.Sample.Middlewares/secureheaders.json]: Take a look at this page please - https://developer.mozilla.org/en-US/docs/Web/HTTP/Public_Key_Pinning 

> The current specification requires including a second pin for a backup key which isn't yet used in production. This allows for changing the server's public key without breaking accessibility for clients that have already noted the pins. This is important for example when the former key gets compromised.

As I understand the second PinSha will be a backup.

[the remark this answers is not in the document]
Comment (rurku) [reply src/WaCore.Sample.Middlewares/secureheaders.json]: Yes that's what I meant. In your example you provide only one key, so browsers would see the header as invalid. I think you should add a second key in the sample.

**Pull request merged**